x-pack/auditbeat/module/system/login/utmp.go

```diff
@@ -0,0 +1,571 @@
+// a corresponding LoginRecord. Some UTMP records do not translate
+// into a LoginRecord, in this case the return value is nil.
+func (r *UtmpFileReader) processLoginRecord(utmp Utmp) *LoginRecord {
+	record := LoginRecord{
+		Utmp:      utmp,
+		Timestamp: utmp.UtTv,
+		UID:       -1,
+		PID:       -1,
+	}
+
+	if utmp.UtLine != "~" {
+		record.TTY = utmp.UtLine
+	}
+
+	switch utmp.UtType {
+	// See utmp(5) for C constants.
+	case C.RUN_LVL: // 1
+		// The runlevel - though a number - is stored as
+		// the ASCII character of that number.
+		runlevel := string(rune(utmp.UtPid))
+
+		// 0 - halt; 6 - reboot
+		if utmp.UtUser == "shutdown" || runlevel == "0" || runlevel == "6" {
+			record.Type = LoginRecordTypeShutdown
+
+			// Clear any old logins
+			// TODO: Issue logout events for login events that are still around
+			// at this point.
+			r.loginSessions = make(map[string]LoginRecord)
+		} else {
+			// Ignore runlevel changes that are not halt or reboot.
+			return nil
+		}
+	case C.BOOT_TIME: // 2
+		if utmp.UtLine == "~" && utmp.UtUser == "reboot" {
+			record.Type = LoginRecordTypeBoot
+
+			// Clear any old logins
+			// TODO: Issue logout events for login events that are still around
+			// at this point.
+			r.loginSessions = make(map[string]LoginRecord)
+		} else {
+			// Ignore unknown record
+			return nil
+		}
+	case C.USER_PROCESS: // 7
+		record.Type = LoginRecordTypeUserLogin
+
+		record.Username = utmp.UtUser
+		record.UID = lookupUsername(record.Username)
+		record.PID = utmp.UtPid
+		record.IP = newIP(utmp.UtAddrV6)
+		record.Hostname = utmp.UtHost
+
+		// Store TTY from user login record for enrichment when user logout
+		// record comes along (which, alas, does not contain the username).
+		r.loginSessions[record.TTY] = record
+	case C.DEAD_PROCESS: // 8
+		savedRecord, found := r.loginSessions[record.TTY]
+		if found {
+			record.Type = LoginRecordTypeUserLogout
+			record.Username = savedRecord.Username
+			record.UID = savedRecord.UID
+			record.PID = savedRecord.PID
+			record.IP = savedRecord.IP
+			record.Hostname = savedRecord.Hostname
+		} else {
+			// Skip - this is usually the DEAD_PROCESS event for
+			// a previous INIT_PROCESS or LOGIN_PROCESS event -
+			// those are ignored - (see default case below).
+			return nil
+		}
+	default:
+		/*
+			Every other record type is ignored:
+			- EMPTY - empty record
+			- NEW_TIME and OLD_TIME - could be useful, but not written when time changes,
+			  at least not using `date`
+			- INIT_PROCESS and LOGIN_PROCESS - written on boot but do not contain any
+			  interesting information
+			- ACCOUNTING - not implemented according to manpage
+		*/
+		return nil
+	}
+
+	return &record
+}
+
+// lookupUsername looks up a username and returns its UID.
+// It does not pass through errors (e.g. when the user is not found)
+// but will return -1 instead.
+func lookupUsername(username string) int {
+	if username != "" {
+		user, err := user.Lookup(username)
```

Comment:
This is probably fine for low volumes. But we have seen from the auditd module that this call is expensive. Some caching could be a useful future enhancement.